### PR TITLE
Fix `jj workspace update-stale` with missing operation and sparse patterns

### PR DESCRIPTION
**Pull request by martinvonz**
# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

File: lib/src/local_working_copy.rs [hunk 1]
```diff
@@ -1457,9 +1457,9 @@ impl TreeState {
         Ok(())
     }
 
-    pub fn reset_to_empty(&mut self) {
+    pub fn recover(&mut self, new_tree_id: MergedTreeId) {
         self.file_states.clear();
-        self.tree_id = self.store.empty_merged_tree_id();
+        self.tree_id = new_tree_id;
```
Comment (yuja) [inline lib/src/local_working_copy.rs:1462]: In `snapshot()`, `deleted_files` are detected by subtracting from the `file_states`. So resetting only the `tree_id` would break the assumption there.

If I changed the test as follows, the debug assertion failed.
```diff
diff --git a/cli/tests/test_workspaces.rs b/cli/tests/test_workspaces.rs
index edb85c84f4...3924bf144d 100644
--- a/cli/tests/test_workspaces.rs
+++ b/cli/tests/test_workspaces.rs
@@ -430,7 +430,8 @@
         &["sparse", "set", "--clear", "--add=file1"],
     );
     // Also leave an unsnapshotted write in the secondary working copy
-    std::fs::write(secondary_path.join("file1"), "secondary\n").unwrap();
+    //std::fs::write(secondary_path.join("file1"), "secondary\n").unwrap();
+    std::fs::remove_file(secondary_path.join("file1")).unwrap();
 
     // Create an op by abandoning the parent commit. Importantly, that commit also
     // changes the target tree in the secondary workspace.
```

Comment (martinvonz) [reply lib/src/local_working_copy.rs]: Good catch! Thanks for finding that. I've added some more tests and updated the code to basically first reset to the empty tree and then reset to the recovery commit.

**Review by yuja — APPROVED** [review]
